File: src/main/java/com/cxxcxx/zinecraft/core/datagen/ZinecraftDataGenerator.java
package com.cxxcxx.zinecraft.core.datagen;

import com.cxxcxx.zinecraft.api.datagen.CatalogLanguageProvider;
import com.cxxcxx.zinecraft.api.datagen.CatalogLootTableProvider;
import com.cxxcxx.zinecraft.api.datagen.CatalogModelProvider;
import com.cxxcxx.zinecraft.api.datagen.RegistryDataContributor;
import com.cxxcxx.zinecraft.core.Zinecraft;
import com.cxxcxx.zinecraft.core.registry.ModRecipeProvider;
import net.minecraft.core.RegistrySetBuilder;
import net.neoforged.neoforge.data.event.GatherDataEvent;

import java.util.List;

/**
 * NeoForge data-generation entrypoint.
 */
public final class ZinecraftDataGenerator {
  private ZinecraftDataGenerator() {
  }

  public static void gatherData(GatherDataEvent event) {
    Zinecraft.bootstrapContent();

    var output = event.getGenerator().getPackOutput();
    var lookup = event.getLookupProvider();
    if (event.includeClient()) {
      event.addProvider(new CatalogLanguageProvider(output, Zinecraft.TRANSLATIONS, "en_us"));
      event.addProvider(new CatalogLanguageProvider(output, Zinecraft.TRANSLATIONS, "zh_cn"));
      event.addProvider(new CatalogModelProvider(output, Zinecraft.ITEMS, Zinecraft.BLOCKS));
    }
    if (event.includeServer()) {
      var registryBuilder = new RegistrySetBuilder();
      List<RegistryDataContributor> contributors = List.of(
          Zinecraft.BIOMES,
          Zinecraft.FEATURES,
          Zinecraft.DENSITY_FUNCTIONS,
          Zinecraft.DIMENSIONS,
          Zinecraft.STRUCTURES,
          Zinecraft.ENCHANTMENTS,
          Zinecraft.SOUNDS
      );
      contributors.forEach(contributor -> contributor.contribute(registryBuilder));
      event.createDatapackRegistryObjects(registryBuilder);

      String terrainHeightmapOutput = System.getProperty(TerraTerrainHeightmapProvider.OUTPUT_PROPERTY);
      if (terrainHeightmapOutput != null && !terrainHeightmapOutput.isBlank()) {
        event.addProvider(new TerraTerrainHeightmapProvider(event.getLookupProvider()));
      }

      event.addProvider(new CatalogLootTableProvider(output, lookup, Zinecraft.BLOCKS, Zinecraft.ENTITIES));
      event.addProvider(new ModRecipeProvider(output, lookup));
    }
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/datagen/ZinecraftDataGenerator.kt
package com.cxxcxx.zinecraft.core.datagen

import com.cxxcxx.zinecraft.api.datagen.CatalogLanguageProvider
import com.cxxcxx.zinecraft.api.datagen.CatalogLootTableProvider
import com.cxxcxx.zinecraft.api.datagen.CatalogModelProvider
import com.cxxcxx.zinecraft.core.Zinecraft
import com.cxxcxx.zinecraft.core.biome.NationBiomes
import com.cxxcxx.zinecraft.core.block.ModBlock
import com.cxxcxx.zinecraft.core.dimension.ModDimensions
import com.cxxcxx.zinecraft.core.entity.ModEntities
import com.cxxcxx.zinecraft.core.item.ModItem
import com.cxxcxx.zinecraft.core.recipe.ModRecipeProvider
import com.cxxcxx.zinecraft.core.skill.ModSkills
import com.cxxcxx.zinecraft.core.sound.ModSound
import com.cxxcxx.zinecraft.core.structure.ModStructure
import com.cxxcxx.zinecraft.core.structure.NationLandmarks
import com.cxxcxx.zinecraft.core.structure.NationSettlements
import com.cxxcxx.zinecraft.core.worldgen.ModWorldFeatures
import com.cxxcxx.zinecraft.core.weapon.ModWeapons
import com.cxxcxx.zinecraft.core.weapon.ModTaczWeapons
import net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint
import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator
import net.minecraft.core.RegistrySetBuilder
import net.minecraft.core.registries.Registries

// fabric 只能有一个数据生成器入口
object ZinecraftDataGenerator : DataGeneratorEntrypoint {
  override fun onInitializeDataGenerator(fabricDataGenerator: FabricDataGenerator) {
    ModSound
    ModItem
    ModBlock
    ModSkills
    ModWeapons
    ModTaczWeapons
    ModEntities
    NationBiomes
    ModDimensions
    NationLandmarks
    NationSettlements
    ModWorldFeatures
    ModStructure
    val pack = fabricDataGenerator.createPack()
    pack.addProvider { output, registries ->
      CatalogLanguageProvider(output, registries, Zinecraft.TRANSLATIONS, "en_us")
    }
    pack.addProvider { output, registries ->
      CatalogLanguageProvider(output, registries, Zinecraft.TRANSLATIONS, "zh_cn")
    }
    pack.addProvider(::ModCatalogModelProvider)
    pack.addProvider { output, registries ->
      CatalogLootTableProvider(output, registries, Zinecraft.BLOCKS)
    }
    pack.addProvider(::ModRecipeProvider)
    pack.addProvider(::ModDynamicRegistryProvider)
  }

  override fun buildRegistry(registryBuilder: RegistrySetBuilder?) {
    registryBuilder?.let(Zinecraft.WORLDGEN::addDataGeneration)
    registryBuilder?.add(Registries.ENCHANTMENT, Zinecraft.ENCHANTMENTS::bootstrap)
    registryBuilder?.add(Registries.JUKEBOX_SONG, ModSound::configure)
  }
}

private class ModCatalogModelProvider(output: net.fabricmc.fabric.api.datagen.v1.FabricDataOutput) :
  CatalogModelProvider(output, Zinecraft.ITEMS, Zinecraft.BLOCKS)
